### ISSUE TITLE
Define redundency on filesystem level

### ISSUE THREAD

**Issue by SSDpreowner** [openzfs/zfs#10972]
### Describe the feature would like to see added to OpenZFS

It should be possible to define a required redundancy level for a filesystem, so the redundant raidz mirror and striped filesystems will be able to share the same disk array.

### How will this feature improve OpenZFS?

It would greatly improve zfs convenience for home usage.
Consider a desktop PC with two identical drives. The user would be able to create a pool over them and then create a mirror filesystem for important files and a striped filesystem for unimportant data that can be redownloaded if needed (games, movies, etc).

### Additional context

It is now possible to partition disks and create separate zpools on each set of partitions effectively sharing the drive array between them but in this case the amount of space for each of them should be decided beforehand which is inconvenient.

**Comment by Ornias1993** [comment]
If you want a new filesystem, I think you should ask it... somewhere, but not here...

What you want is a complete refactor/redesign of ZFS, even more so than draid.
You can already partition disks and create 2 pools: one for all the striped data and one for all the mirrored data.

**Comment by SSDpreowner** [comment]
@Ornias1993 filesystems already have a similar property `copies`, and I want it to actually guarantee fault tolerance.

I know that I can partition disks and create two separate pools, I've described this inconvinient workaround in the additional context section, the size of each one of them should be determined beforehand. I want to actually share the disk array between filesystems and allocate space on demand in a natural way. I believe the requested feature matches the spirit of zfs which is already able to share the same pool between different filesystems. You can define encryption, block size, compression and many other properties on the filesystem level, but not the desired fault tolerance.

**Comment by Ornias1993** [comment]
@SSDpreowner "filesystems" don't exist, I think you mean "datasets".
Datasets are not "disk aware", even the pools aren't fully "disk aware", they just spread load over vdevs, which have specific rules (one of which is a mirror). 

" I want to actually share the disk array between datasets and allocate space on demand in a natural way"
I know you want to, but you can't. Not by ZFS lacking a feature, but you would need the total structure of ZFS do be rewrithen to do what you want it to do.
(basically create a new filesystem)

"share the same pool between different filesystems"
You don't share the pool between filesystems, you just have multiple datasets per pool of disks. datasets are not in anyway related to redundancy and are NOT redundancy aware at all.

"You can define encryption, block size, compression and many other properties on the filesystem level, but not the desired fault tolerance."
A TLDR about the order of how settings are applied to a writen block:
Create a block -> Apply dataset settings (compression, encryption etc.) -> send to pool -> spread over vdevs -> apply vdev redundancy -> Write to disk

Applying redundancy is quite far away from the dataset settings being applied, it would mean a whole different way of handling data. Aka a whole new filesystem.

**Comment by SSDpreowner** [comment]
@Ornias1993 isn't filesystem a type of datasets, the filesystem term is used all over the zfs manpage, but ok... what about the `copies` property? It is defined on dataset level and according to the manual `copies are stored on different disks, if possible`. How it is possible if it is so far away from the vdev level? Why is it so difficult to make its behaviour "guaranteed" instead of just "if possible"?

**Comment by Ornias1993** [comment]
I'm not really into the nuances, but filesystems not a different word for dataset or zvol.

It can NOT be guaranteed, your vdevs are not guaranteed to have similair freespace. If you have 2 vdevs and one if completely full (for example: due to expansion), it's all writen to the second.

It's also still the same vdevs.
Simply put: you write two copies to raidz-2 devs, it's still raid-z2, it isn't a mirror, because that setting is not disk aware. A vdev with raidz2 is always writhing raidz2.

**Comment by SSDpreowner** [comment]
> If you have 2 vdevs and one is completely full  

And what would it do if both of them are full and neither of the devices can store two copies of data? Does pool return back error in this case? Why can't it return error if the copies can't be evenly mirrored between disks then? If the dataset level is aware of the underlying disk space why can't it decide that it is not enough space remained to store mirrored copy and return error to the caller?

**Comment by Ornias1993** [comment]
Thats out-of-scope and irrelevant to your feature request.

Because what you just described (guaranteed spreading copies) still is NOT a mirrored pool. It has all the downsides of a mirror AND the underlaying vdevs.

**Comment by SSDpreowner** [comment]
> guaranteed spreading copies still is NOT a mirrored pool

If two copies are evenly spread between disks it would be effectively a mirror for two vdevs. I will be able to restore the data if one vdev fails. Isn't it the goal of the copies property? It is just not guaranteed now.

I've described my usecase in the feature request. I have two disks, I want to put them in a single pool and put some unimportant data like games in one filesystem/dataset  and store some important data in a different dataset. Basically what I want now is to make sure that copies are guaranteed. The feature request is formulated more broadly but still. This is what I want:

```
zpool create tank /dev/sda /dev/sdb
zfs create tank/games
zfs create -o redundency=mirror tank/documents
```
I can write "zfs create -o copies=1 tank/documents" but its fault tolerance is not guaranteed.

Well. I am the user, I've described the possible improvement in the request. From my user point of view it feels natural and convenient to define required redundancy as a dataset property. You are developer you know better how difficult is to implement it. If you think it wouldn't be implemented anytime soon, this is sad.

**Comment by Ornias1993** [comment]
Redundancy is a per-vdev setting and is not ever going to be per-dataset. thats basically it.

> I will be able to restore the data if one vdev fails

Yes and no, you are not going to be able to restore the pool afaik.
You might be able to get some data from it with some tricks and save the data, but that would take a significant amount of work.

**Comment by SSDpreowner** [comment]
> Your proposal is very close to "how not to use ZFS".

My proposal is to implement a guaranteed redundancy feature on dataset level. I am not going to use copies property workaround while it is not guaranteed.

The only ugly but currently available workaround to partition drives and make two pools of a fixed size over them is indeed "how not to use ZFS", I've heard that sharing the same drive between pools this way is not recommended.

**Comment by Ornias1993** [comment]
Copies != redundancy
It's (among other things) a protection against bitrot, not disk failure!

You want to use copies as redundancy, it isn't. Not even while guaranteed to be spread. 
Your pool can still fail completely if one disk fails and you couldn't repair the pool afaik. It might(!) just be possible to do some form of data recovery.

**Comment by mschilli87** [comment]
@SSDpreowner:
> If two copies are evenly spread between disks it would be effectively a mirror for two vdevs.

I think this is far from true.
You can have the same data on two disks without these disks being mirrors.

Consider the following though experiment:
There are three pairs of (pairwise) identical coins on the table.
I pick one each of each pair, put two of them in one of my pockets and the third in the other pocket without you being able to see it.
Now you enter the room, pick up the remaining three coins and put two of them in one of your pockets and the third in your other pocket as well.
Accoring to you reasong, we would no be guaranteed to have the same combination of two coins in one pocket (and thus the same coin on its own in the other). Furthermore, we would both have the single coin in the same pocket (right/left).
If I stand in front of a mirror, these things hold true for me an my mirror image.
For you and me, all we have in common is that we hold the same coins (data) in a similar fashion (layout). But we do *not* mirror each other.

**Comment by Ornias1993** [comment]
@mschilli87 Correct me if i'm wrong but about that manual portion:

> copies are stored on different disks, if possible

Shouldn't that be

> copies are stored on different vdevs, if possible

**Comment by mschilli87** [comment]
@Ornias1993:

> > copies are stored on different disks, if possible
> 
> Shouldn't that be
> 
> > copies are stored on different vdevs, if possible

My gut feeling tells me yes, but I honestly don't know for sure.

**Comment by SSDpreowner** [comment]
**@Ornias1993:**
> You want to use copies as redundancy, it isn't

I didn't say otherwise. I've only bring copies property as an example of a dataset level property that influence the way blocks are stored on vdevs "if possible". It is not "guaranteed" and not redundant. I've repeated this in every comment here.

**mschilli87:**
>  IMHO there is no cheap shortcut

It would be a cheap shortcut if only it was possible to define required redundancy level for each dataset separately and share the available disk array space between them. Hence the feature request.

**Comment by Ornias1993** [comment]
> I didn't say otherwise. I've only bring copies property as an example of a dataset level property that influence the way blocks are stored on vdevs "if possible". It is not "guaranteed" and not redundant. I've repeated this in every comment here.

Thats not what i'm saying, even IF it was guaranteed, it still wouldn't be redundancy. Both serve a different purpose. Copies (even when spread out correctly) DO NOT save your pool on disk failure.

**Comment by shodanshok** [comment]
@SSDpreowner as far I know, what you are asking can not be done inside the current ZFS framework. The key point is that block allocation is done at the SPA (ie: pool) level, while dataset don't know anything about how physical block are allocated. Even the `copies` property does not guarantee that multiple copies are really allocated on different disks, at a point that ZFS simply does not import a striped vdev with a missing disk even if `copies=3` *and all data/metadata were really stored on different disks*
I would love to be proven wrong, but I think we will never see such feature in ZFS...

**Comment by Ornias1993** [comment]
@shodanshok Any idea about that thing about seperating the copies over different drivers thats in the manual?

I really think thats wrong, if it is done it should be spread about vdev's not disks. is it actually doing so, or it just happens to often be spread over the drivers?

**Comment by shodanshok** [comment]
@Ornias1993 I think you are right: ditto blocks are spread amont top-level vdev rather than physical disks. Consider, for example, such a test pool:

```
[root@locahost shm]# zpool status zzz
  pool: zzz
 state: ONLINE
  scan: none requested
config:

        NAME                    STATE     READ WRITE CKSUM
        zzz                     ONLINE       0     0     0
          mirror-0              ONLINE       0     0     0
            /dev/shm/disk1.img  ONLINE       0     0     0
            /dev/shm/disk2.img  ONLINE       0     0     0
            /dev/shm/disk3.img  ONLINE       0     0     0
            /dev/shm/disk4.img  ONLINE       0     0     0
          mirror-1              ONLINE       0     0     0
            /dev/shm/disk5.img  ONLINE       0     0     0
            /dev/shm/disk6.img  ONLINE       0     0     0
            /dev/shm/disk7.img  ONLINE       0     0     0
            /dev/shm/disk8.img  ONLINE       0     0     0

errors: No known data errors
```

As you can see, it has 8 disks and 2 top-level vdev. I created a test dataset called `zzz/test` with `copies=3`, then I wrote 32K of random data. `zdb -ddddd -bbbbbb zzz/test 2` shows the following:

```
Dataset zzz/test [ZPL], ID 131, cr_txg 10, 135K, 9 objects, rootbp DVA[0]=<1:24a00:200> DVA[1]=<0:24a00:200> DVA[2]=<1:101fc00:200> [L0 DMU objset] fletcher4 lz4 unencrypted LE contiguous unique triple size=1000L/200P birth=11L/11P fill=9 cksum=11785cb1e3:65b7f098aea:1364cc4bd22a9:292c053f31d440

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    1   128K    32K    96K     512    32K  100.00  ZFS plain file
                                               184   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED
        dnode maxblkid: 0
        path    /test.img
        uid     0
        gid     0
        atime   Sat Sep 26 12:22:59 2020
        mtime   Sat Sep 26 12:22:59 2020
        ctime   Sat Sep 26 12:22:59 2020
        crtime  Sat Sep 26 12:22:59 2020
        gen     11
        mode    100644
        size    32768
        parent  34
        links   1
        pflags  840800000004
        xattr   3
Indirect blocks:
               0 L0 DVA[0]=<1:16800:8000> DVA[1]=<0:16800:8000> DVA[2]=<1:1011a00:8000> [L0 ZFS plain file] fletcher4 uncompressed unencrypted LE contiguous unique triple size=8000L/8000P birth=11L/11P fill=1 cksum=ff4106204d6:1000da0ddc07ff2:ad4730f415401c89:92f0bc12d80b2fc

                segment [0000000000000000, 0000000000008000) size   32K
```

Based on the line with DVA info, the three ditto blocks were located on vdev 1, 0 and 1. The fact that each vdev is a 4-way mirror is an underlying property of the vdev itself, rathen than being controlled by the `copies` property.

I think the confusion regarding the man page is that the `copies` property is mostly useful with single-disk vdevs - ie: a pool similar to that: 

```
[root@nas shm]# zpool status zzz
  pool: zzz
 state: ONLINE
  scan: none requested
config:

        NAME                  STATE     READ WRITE CKSUM
        zzz                   ONLINE       0     0     0
          /dev/shm/disk1.img  ONLINE       0     0     0
          /dev/shm/disk2.img  ONLINE       0     0     0
          /dev/shm/disk3.img  ONLINE       0     0     0
          /dev/shm/disk4.img  ONLINE       0     0     0
          /dev/shm/disk5.img  ONLINE       0     0     0
          /dev/shm/disk6.img  ONLINE       0     0     0
          /dev/shm/disk7.img  ONLINE       0     0     0
          /dev/shm/disk8.img  ONLINE       0     0     0

errors: No known data errors
```

So no redundancy at all. On this pool, `zdb -ddddd -bbbbbb zzz/test 2` returns:

```
Dataset zzz/test [ZPL], ID 166, cr_txg 36, 135K, 9 objects, rootbp DVA[0]=<1:25000:200> DVA[1]=<2:fa00:200> DVA[2]=<3:d200:200> [L0 DMU objset] fletcher4 lz4 unencrypted LE contiguous unique triple size=1000L/200P birth=37L/37P fill=9 cksum=f49d32d71:58c49a09a78:10e188181a712:23c4c57f13b84a

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    1   128K    32K    96K     512    32K  100.00  ZFS plain file
                                               184   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED
        dnode maxblkid: 0
        path    /test.img
        uid     0
        gid     0
        atime   Sat Sep 26 12:21:25 2020
        mtime   Sat Sep 26 12:21:25 2020
        ctime   Sat Sep 26 12:21:25 2020
        crtime  Sat Sep 26 12:21:25 2020
        gen     37
        mode    100644
        size    32768
        parent  34
        links   1
        pflags  840800000004
        xattr   3
Indirect blocks:
               0 L0 DVA[0]=<1:2a00:8000> DVA[1]=<2:3a00:8000> DVA[2]=<3:1200:8000> [L0 ZFS plain file] fletcher4 uncompressed unencrypted LE contiguous unique triple size=8000L/8000P birth=37L/37P fill=1 cksum=101ee0c1912c:1034eeb5cbaf369:d707e4e629f4adad:e574ef6cf5abefd7

                segment [0000000000000000, 0000000000008000) size   32K

```

Data and metadata are now spread on different disk, but only because in such a pool *the disk and vdev are the same, having a 1-to-1 relationship*.

This matches my current understanding of DMU and SPA layer, but I can be wrong of course.

**Comment by Ornias1993** [comment]
@shodanshok Precisely, I'll look into updating the manual later this week.  :)

**Comment by jittygitty** [comment]
@SSDpreowner Great feature request even if final verdict might be it can't be done with ZFS. I hate to send you to BTRFS but it might have the ability already to do what you want, or if not it seems at least their architecture should make it very easy to implement. Take a look at:

"File Striping, File Mirroring, File Striping+Mirroring"
https://btrfs.wiki.kernel.org/index.php/Main_Page

I've used/followed ZFS for some time but been too busy to interact.  I will provide some quick advice because I really wish for the Long-Term success of ZFS and am very thankful for the hard work of those who contributed. 

So again, please don't take this the wrong way, I really wish for ZFS to outlive all other filesystems. Even love the name, butter-fs not so much! 

1. ZFS will die sooner, if the developers are unable to address more "regular" users' concerns. Having a few companies to pay for ZFS development is great, but these companies are not as stable as regular "consumers". NTFS has been around as long as it has because "everyone" used NTFS. ZFS as well as many other open source projects have neglected to use the power of the "individuals"/consumers and have relied on company sponsors.

2. If ZFS project opens itself up to proactively be more receptive of "regular user"/consumer feature requests, mark my words that some great things will happen. Give users a chance to "CLICK" and Donate on a FEATURE. Your company doesn't need it so you don't want it to be worked on? You're afraid a lot of money might be raised to work on that feature and you're jealous it can't be you but somebody else? What's the reasoning? 

I have seen countless open source projects suffer and eventually die because developers refused to listen to the "regular" users. Usually its because the developers were only interested in doing what they wanted and to them listening to "users" meant they had a "boss" again telling them what to do, but they wanted to be the only boss.

Hasn't "crowd funding" been already proven a success? Isn't it in the spirit of community open source?

3. Let's give a concrete example of a request every consumer has been asking/begging for...a long time:

BTRFS equivalent REFLINK support, ia the cp --reflink feature that I've heard Oracle ZFS has implemented.

Let me ask the experts a question here if I may, would it be possible to be able to in a sense "rebase" after changing the "source" of reflinks? My idea is what if you have a base system installed and you did cp --reflinks of the original/source virtual machine image and later you want to update some "drivers". The dream would be to update the drivers on the "source" image and then run some "rebase" which would trickle down the changes to all the reflinks and bingo you just upgraded the drivers on all the reflinked virtual machines, would be nice.

Anyway simply having some sort of "Offline Deduplication" would be a great thing and would increase zfs usage by regular users on their own Linux systems and Mac OS X and hopefully even Windows in near future. Its a bad idea to tell users to go find their own deduplication tools, even BTRFS is making that mistake which is puzzling since their system has gone through the effort to put in place everything to make it work but stopped short of integrating an actual official tool to do what they advertise as their feature! We need to include important tools.

As an example say you run zfs-dedupe and you get asked if you want to use hard-links first when that's feasible, then second it might ask you if to do block-level etc offline-dedupe and/or if to enable live dedupe on smaller datasets. ZFS should 'include' tools necessary to make "good use" of its features and not rely on users testing out dozens of tools and all deciding on a different one. Variety is great but not a substitute for a default tool.

Currently I believe that Microsoft Windows Server has the BEST Offline Deduplication (better than BTRFS), looks like everyone loves it, not sure if its tied to their new ReFS filesystem or works on NTFS also. This is important not just for backup machines to deduplicate operating system files etc but for users who dump multiple copies of their thousands of photos etc and don't want to bother cleaning it up manually and similar issues. Company servers don't suffer from such problems because data is more organized and thought out, end users must have tools to make up for their laziness and "lack of time". And in the long-term its worth pandering to "consumers".

Of course live real-time Deduplication like ZFS had in mind was ahead of its time and unfortunately just about nobody uses due to some issues. If that can be improved perhaps that's the better avenue, you'd know better.

Speaking of tools, there should be a tool/script that "fixes" the KERNEL SOURCES we've downloaded to reverse the source code changes and to EXPORT all the GPL SYMBOLS. Once we've downloaded the sources there's no reason we can't do that, why bother with messier not as good workarounds when we have the kernel sources?

Now back again to BTRFS and @SSDpreowner 's original question, how difficult would it be for ZFS to implement a sort of FILE LEVEL mirroring? Btrfs can do some things on a "directory" level or even a "file" level.

I'm guessing ZFS architecture is much different and it can't have such granularity and flexibility as BTRFS, right?

Another thing that would be nice to get some fixes if possible is the usage of ZVOL's, since some have written about major issues with snapshots on ZVOLs and its actually a wonderful feature that makes ZFS really stand out above other file systems, a very versatile feature which I'd love to use more if worked as well as datasets.

But at least for those aspects that aren't as impossible, it would be nice if the OpenZFS project leadership would make a little more effort in facilitating the implementation of features that regular users want and not just those of the "sponsoring companies". Users want to be "sponsors" too, but they want accountability same as those companies, we don't want to "blindly donate", we want a chance to fund/sponsor specific FEATURES we need : )

There were times I would have clicked to commit to few thousand dollars to ZFS, or other project, but it seemed like user funding was not welcome because regular user features were not truly welcome, unless they were on the "todo/need" of the sponsoring "companies". I bet "many" other would-be donors were in same boat as me!
We want to fund, but by "results" kickstarter-style, not just oh please donate, and we'll try/maybe work on that.

And yea would be nice if @behlendorf  @ahrens  would read the above. Thanks to @ryao for prior efforts in pushing some of these issues. And thanks to all developers, lastly but not least big thanks to @tcaputi for bringing us encryption and making it harder for oracle to cherry-pick from our tree without giving back : ) 

(Oh and a shout out to those who did the aes/gcm contributions/assembly etc in improving performance!)
(And I'm sure many consumers also appreciate the recent efforts on the vdev config change flexibility thx!)

Sorry for the long 1st post, but rest assured I've read a WHOLE LOT MORE of what all you guys have written : )

**Comment by Ornias1993** [comment]
@jittygitty Dude, don't hijack someone elses topic with offtopic blablabla.
Go to the mailinglist. period.

Besides, 1. clearly shows you have no idea what you are talking about. Because A LOT of us are here on a personal capacity (too), not because our boss wants us to. We all donate A LOT to this project that isn't because of a company.

That being said: ZFS is not a "normal user" filesystem like NTFS in the first place.

**Comment by SSDpreowner** [comment]
> @jittygitty Write-a-new-filesystem hard. Because vdevs aren't file-aware as previously explained

The feature doesn't require vdevs be file-aware actually. It only require to pass some info from the dataset level to the pool level to turn mirroring on and off for individual blocks or so. It may require lots of coding and refactoring, it may even require extensions to the ondisk format (which I doubt but I am not competent to say). But it definitely not the new-filesystem-from-scratch as you insisting. ZFS have many many MANY more features than just redundant block storage and this enhancement doesn't require to rewrite any of them. I believe that persistent L2ARC also required lots of coding and changes, now imagine if someone have come to the persistent L2ARC feature request and wrote "If you want a new filesystem with a permanent cache, I think you should ask it... somewhere, but not here...".

Lets just agree that the feature I ask for is very difficult to implement and it is not very awaited to beg for such efforts.

**Comment by Ornias1993** [comment]
@SSDpreowner 
This feature-without-sane-usecase has turned into some sort of community hate-group/cryout. No mater how critical I am about the leadership here, thats not something I condone just based on "we don't get what we want".
I feel continuing this discussion any further would add value to said viewpoint, which is not something i'm interested in.
(the above is totally not your faulth though)

So i'll leave it at "we agree to disagree quite vocally" ;-)

**Comment by fredcooke** [comment]
That's a lot of arguing for a simple feature request. Letting it lie here ignored after the first few non-confrontational replies would have been a good thing. Genuine suggestion for the OP, @SSDpreowner how about this:

1. Format both drives with a simple/efficient file system of your choice, eg encrypted ext4 for me
2. Create on both drives 2 drive-sized sparse files named as you choose
3. With one pair of sparse files create a mirror vdev/zpool
4. With the other pair of sparse files create a raidz vdev/zpool
5. Mount/feature-set/write data to your two pools as desired

Data written to the mirror zpool will utilise the underlying disks almost exactly as you'd expect (1x write, 2x read, 2x redundancy). Ditto the raidz (2x write, 2x read, 1x redundancy). Same benefits and drawbacks. 

When the underlying disks are full you'll get an error, but a nightly cron to check the df -h output of the ext4 filesystems should be sufficient to avoid that scenario. My 2c to meet/solve your requirement/problem. 

PS, found this ticket while searching open issues for something I want and am willing to do some leg work on if someone's willing to help me a bit :-) Not found yet, new ticket soon, perhaps.

**Comment by GregorKopka** [comment]
> It may require lots of coding and refactoring, it may even require extensions to the ondisk format (which I doubt but I am not competent to say).

Yes and yes. 
Bottom line: it would be a new file system.

From that: this will likely not happen, unless you do it yourself or _force others with big piles of money_ to do that rewrite.

**Comment by GregorKopka** [comment]
> There were times I would have clicked to commit to few thousand dollars to ZFS, or other project,

You can always offer a bounty for the development of a feature you would like to see - or hire someone ;)
